common/io/Serial.cpp

```diff
@@ -285,19 +271,55 @@ void ReleaseUUCPLock(const std::string &path) {
 }
 
 bool AcquireLockAndOpenSerialPort(const std::string &path, int oflag, int *fd) {
+  // First, check if the path exists, there's no point trying to open it if not
+  if (!FileExists(path)) {
+    OLA_INFO << "Device " << path << " doesn't exist, so there's no point "
+                "trying to acquire a lock";
+    return false;
+  }
+
+#ifdef UUCP_LOCKING
+  if (!AcquireUUCPLock(path)) {
+    return false;
+  }
+#endif
+
+  // Now try to open the serial device.
+  if (!TryOpen(path, oflag, fd)) {
 #ifdef UUCP_LOCKING
-  return AcquireUUCPLockAndOpen(path, oflag, fd);
+    OLA_DEBUG << "Failed to open device " << path << " despite having the "
+              << "lock file";
+    ReleaseUUCPLock(path);
 #else
-  return OpenAndFlock(path, oflag, fd);
+    OLA_DEBUG << "Failed to open device " << path;
 #endif
+    return false;
+  }
+
+#ifdef HAVE_FLOCK
+  if (flock(*fd, LOCK_EX | LOCK_NB) == -1) {
+    OLA_INFO << "Failed to flock() device " << path;
+    close(*fd);
+    return false;
+  } else {
+    OLA_INFO << "Locked " << path << " using flock()";
+  }
+#endif
+
+  if (!LockTIOCEXCL(*fd, path)) {
+    close(*fd);
+    return false;
+  }
+
+  return true;
 }
 
 void ReleaseSerialPortLock(const std::string &path) {
 #ifdef UUCP_LOCKING
   ReleaseUUCPLock(path);
-#else   // UUCP_LOCKING
-  OLA_INFO << "No unlock necessary for " << path;
-#endif  // UUCP_LOCKING
+#else
+  OLA_INFO << "No unlock necessary for " << path << " (UUCP locking not used)";
+#endif
```

Comment:
I think we do want the comment here to keep cpplint happy, just not on the else:
```suggestion
#endif  // UUCP_LOCKING
```